### PR TITLE
Support home channel change

### PR DESCRIPTION
**Pull request by MartinBroers**
Commit to support WIFI_EVENT_HOME_CHANNEL_CHANGE, introduced in https://github.com/espressif/esp-idf/commit/1427e14b724292ee8ceede6523d98cdee7c06037. I have opted to 'just' support the event and not propagate the values which come into the struct in the C library, but I could add those if required.

Also, I guarded it with a config directive (`#[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]`, so not to break backwards builds. Maybe this is too restrictive, and we need to adjust for future builds (anything greater than 5.2), so  I am open to suggestions here. Or is the idea the next release works for 5.2 and not for older versions?

Also, I took the liberty to fix a few compilation errors and added 5.2 to the github workflow, but I am not sure this all is correct.

### CODE REVIEW

File: src/bt/a2dp.rs [hunk 1]
```diff
@@ -134,6 +134,13 @@ pub enum ConnectionStatus {
 #[repr(u32)]
 pub enum AudioStatus {
     SuspendedByRemote = esp_a2d_audio_state_t_ESP_A2D_AUDIO_STATE_REMOTE_SUSPEND,
+    #[cfg(any(
+        esp_idf_version_major = "4",
+        all(esp_idf_version_major = "5", esp_idf_version_minor = "0"),
```
Comment (ivmarkov) [inline src/bt/a2dp.rs:139]: Nit: simplify this by replacing with `all(esp_idf_version_major = "5", any(esp_idf_version_minor = "0", esp_idf_version_minor = "1"))`

File: src/bt/gap.rs [hunk 2]
```diff
@@ -3,6 +3,13 @@
 
 use core::cmp::min;
 use core::convert::TryInto;
+#[cfg(any(
+    esp_idf_version_major = "4",
+    all(esp_idf_version_major = "5", esp_idf_version_minor = "0"),
```
Comment (ivmarkov) [inline src/bt/gap.rs:8]: Ditto.

File: src/eth.rs [hunk 3]
```diff
@@ -462,6 +462,8 @@ where
                     spi_host_id: host,
                     spi_devcfg: &spi_devcfg as *const _ as *mut _,
                     int_gpio_num: int,
+                    #[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Comment (ivmarkov) [inline src/eth.rs:465]: This is not OK, as we'll have to touch it again once 5.3 or 6.0 is out. Rather, replace it with `not(<the upper condition>)`.

File: src/eth.rs [hunk 4]
```diff
@@ -490,6 +492,8 @@ where
...
```
Comment (ivmarkov) [inline src/eth.rs:495]: Ditto.

File: src/eth.rs [hunk 5]
```diff
@@ -518,6 +522,8 @@ where
...
```
Comment (ivmarkov) [inline src/eth.rs:525]: Ditto.

File: src/wifi.rs [hunk 6]
```diff
@@ -1888,67 +1924,65 @@ impl<'a> EspEventDeserializer for WifiEvent<'a> {
...
+            wifi_event_t_WIFI_EVENT_AP_STOP => WifiEvent::ApStopped,
+            wifi_event_t_WIFI_EVENT_AP_STACONNECTED => WifiEvent::ApStaConnected,
+            wifi_event_t_WIFI_EVENT_AP_STADISCONNECTED => WifiEvent::ApStaDisconnected,
+            wifi_event_t_WIFI_EVENT_AP_PROBEREQRECVED => WifiEvent::ApProbeRequestReceived,
+            wifi_event_t_WIFI_EVENT_FTM_REPORT => WifiEvent::FtmReport,
+            wifi_event_t_WIFI_EVENT_STA_BSS_RSSI_LOW => WifiEvent::StaBssRssiLow,
+            wifi_event_t_WIFI_EVENT_ACTION_TX_STATUS => WifiEvent::ActionTxStatus,
+            wifi_event_t_WIFI_EVENT_STA_BEACON_TIMEOUT => WifiEvent::StaBeaconTimeout,
+            wifi_event_t_WIFI_EVENT_ROC_DONE => WifiEvent::RocDone,
+            #[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Comment (ivmarkov) [inline src/wifi.rs:1970]: Ditto - see above.

File: src/wifi.rs [hunk 7]
```diff
@@ -1800,6 +1800,39 @@ impl<'d> NetifStatus for EspWifi<'d> {
 #[repr(transparent)]
 pub struct WpsCredentialsRef(wifi_event_sta_wps_er_success_t__bindgen_ty_1);
 
+#[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Comment (ivmarkov) [inline src/wifi.rs:1803]: Ditto - see above.

File: src/wifi.rs [hunk 8]
```diff
@@ -1800,6 +1800,39 @@ impl<'d> NetifStatus for EspWifi<'d> {
...
+#[derive(Copy, Clone, Debug)]
+#[repr(C)]
+pub struct HomeChannelChange {
+    old_chan: u8,
+    old_snd: WifiSecondChan,
+    new_chan: u8,
+    new_snd: WifiSecondChan,
+}
+
+#[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Comment (ivmarkov) [inline src/wifi.rs:1813]: Ditto - see above.

File: src/wifi.rs [hunk 9]
```diff
@@ -1800,6 +1800,39 @@ impl<'d> NetifStatus for EspWifi<'d> {
...
+#[derive(Copy, Clone, Debug)]
+enum WifiSecondChan {
+    None = 0,
+    Above,
+    Below,
+}
+
+#[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Comment (ivmarkov) [inline src/wifi.rs:1821]: Ditto - see above.

File: src/wifi.rs [hunk 10]
```diff
@@ -1873,6 +1906,9 @@ pub enum WifiEvent<'a> {
     FtmReport,
     ActionTxStatus,
     RocDone,
+
+    #[cfg(all(esp_idf_version_major = "5", esp_idf_version_minor = "2",))]
```
Comment (ivmarkov) [inline src/wifi.rs:1910]: Ditto - see above.

**Review by ivmarkov — CHANGES_REQUESTED** [review]
I think it is in a very good shape. However, if you could fix the `#cfg`s - see my other comments.

**Comment by MartinBroers** [comment]
Thanks @ivmarkov , some good suggestions.

**Comment by MartinBroers** [comment]
Updated MR to 
- Compile again. Something in rust-nightly has changed causing clippy errors. Fixed those in the last commit
- Not use BT in the sample sdkconfig, so we don't need the update on esp-idf-sys.

File: src/http/client.rs [hunk 11]
```diff
@@ -9,10 +9,9 @@ use core::cell::UnsafeCell;
 use core::fmt::Write as _;
 
 extern crate alloc;
-use alloc::boxed::Box;
 use alloc::collections::BTreeMap;
-use alloc::string::String;
-use alloc::string::ToString;
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/http/client.rs:13]: Should be `feature = "std"`, NOT `feature = "default"`.

File: src/http/server.rs [hunk 12]
```diff
@@ -40,13 +40,11 @@ use core::time::*;
 use core::{ffi, ptr};
 
 extern crate alloc;
-use alloc::borrow::ToOwned;
-use alloc::boxed::Box;
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/http/server.rs:43]: Ditto.

File: src/mdns.rs [hunk 13]
```diff
@@ -3,9 +3,8 @@
 use core::time::Duration;
 
 extern crate alloc;
-use alloc::string::String;
-use alloc::string::ToString;
-use alloc::vec::Vec;
+#[cfg(not(feature = "default"))]
+use alloc::{string::String, string::ToString, vec::Vec};
```
Comment (ivmarkov) [inline src/mdns.rs:7]: Ditto.

File: src/mqtt/client.rs [hunk 14]
```diff
@@ -1,10 +1,10 @@
 //! MQTT protocol client
-use core::convert::{TryFrom, TryInto};
 use core::ffi::c_void;
 use core::fmt::Debug;
 use core::{slice, time};
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/mqtt/client.rs:7]: Ditto

File: src/timer.rs [hunk 15]
```diff
@@ -10,11 +10,11 @@
 //! microsecond time resolution, and 52-bit range.
 
 use core::num::NonZeroU32;
-use core::result::Result;
 use core::time::Duration;
 use core::{ffi, ptr};
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/timer.rs:17]: Ditto

File: src/wifi.rs [hunk 16]
```diff
@@ -1,11 +1,11 @@
 //! WiFi support
-use core::convert::{TryFrom, TryInto};
 use core::marker::PhantomData;
 use core::str::Utf8Error;
 use core::time::Duration;
 use core::{cmp, ffi, fmt};
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/wifi.rs:8]: Ditto.

File: src/ws/client.rs [hunk 17]
```diff
@@ -1,10 +1,13 @@
 //! WebSocket client
 
-use core::convert::{TryFrom, TryInto};
 use core::{ffi, time};
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/ws/client.rs:4]: Ditto.

File: .github/configs/sdkconfig.defaults [hunk 18]
```diff
@@ -18,17 +18,7 @@ CONFIG_ETH_SPI_ETHERNET_W5500=y
...
-CONFIG_BT_CLASSIC_ENABLED=y
-CONFIG_BTDM_CTRL_MODE_BLE_ONLY=n
-CONFIG_BTDM_CTRL_MODE_BR_EDR_ONLY=y
-CONFIG_BTDM_CTRL_MODE_BTDM=n
-CONFIG_BT_A2DP_ENABLE=y
-CONFIG_BT_HFP_ENABLE=y
-CONFIG_BT_HFP_CLIENT_ENABLE=y
-CONFIG_BT_HFP_AUDIO_DATA_PATH_HCI=y
-CONFIG_BT_BLE_42_FEATURES_SUPPORTED=y
+CONFIG_BT_ENABLED=n
```
Comment (ivmarkov) [inline .github/configs/sdkconfig.defaults:21]: Please restore the BT settings, but commented out, and put a short comment on top: "Uncomment to enable Classic BT"

File: src/espnow.rs [hunk 19]
```diff
@@ -10,6 +10,7 @@ use core::marker::PhantomData;
 
 use ::log::info;
 
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/espnow.rs:13]: See below.

File: src/eth.rs [hunk 20]
```diff
@@ -6,6 +6,7 @@ use core::{ffi, ptr};
 use ::log::*;
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/eth.rs:9]: See below.

File: src/eventloop.rs [hunk 21]
```diff
@@ -1,13 +1,12 @@
...
 
-use core::convert::{TryFrom, TryInto};
 use core::fmt::Debug;
 use core::marker::PhantomData;
-use core::result::Result;
 use core::time::Duration;
 use core::{ffi, mem, ptr, slice};
 
 extern crate alloc;
+#[cfg(not(feature = "default"))]
```
Comment (ivmarkov) [inline src/eventloop.rs:9]: See below.

**Review by ivmarkov — CHANGES_REQUESTED** [review]

**Comment by ivmarkov** [comment]
@MartinBroers W.r.t. the clippy errors... I'm thinking we need a more generic fix. As in - for example - not importing the `std` prelude - ever - even if the `std` feature is enabled. Otherwise we need to sprincle the whole code with conditional imports, which is not nice at all.

**Pull request merged**